### PR TITLE
Week6

### PR DESCRIPTION
**Pull request by PotzarellaMozarella**
Week 6 Assignments:

1. Max Sum Contiguous Subarray:
Find the contiguous subarray within an array, A of length N which has the largest sum.Input Format:The first and the only argument contains an integer array, A. Output Format: Return an integer representing the maximum possible sum of the contiguous subarray.

2. Spiral Order Matrix II
Problem Description Given a matrix of m * n elements (m rows, n columns), return all elements of the matrix in spiral order.

3. Sort array of 0's,1's and 2's
Given an array of size N containing only 0s, 1s, and 2s; sort the array in ascending order.

4. Best time to buy and sell stock
You are given an array prices where prices[i] is the price of a given stock on the ith dayYou want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock. Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.

6. Pair With Given Difference
Given a one-dimensional unsorted array A containing N integers. You are also given an integer B, find if there exists a pair of elements in the array whose difference is B.

7. 3 sum
Given an array S of n integers, find three integers in S such that the sum is closest to a given number, target. Return the sum of the three integers. Assume that there will only be one solution.

### CODE REVIEW

File: Week-3/Assignment1/main.js [hunk 1]
```diff
@@ -1,44 +1,50 @@
-
 //reducer method
```
Comment (imranmd) [inline Week-3/Assignment1/main.js:1]: Week6 review:

1. Give meaningful names to variables and file names, example in Q1, you passed 'A', instead give some meaningful param name
2. Edge cases such as null or undefined is not covered. For example If you pass maxSubArray(null/undefined), your code will break
3. Logic seems to be wrong for Assignment 1, Please revisit question
4. Good work on assignment 3, Did you evaluate any other approaches, If yes why this is optimal, You can avoid some extra space used here.
5. Awesome work on spiral matrix, Keep it up, Try to add in comments if there are any other alternative approaches you thought of.
6. Assignment 5, second test case is failing, Please check, I guess you are missing modulus here.

Comment (PotzarellaMozarella) [reply Week-3/Assignment1/main.js:1]: 1. Done
2. Done
3. Checked the logic with multiple outputs, can you let me know which part of the logic is wrong. Thanks!
4. Updated another approach and revised the existing approach as well to optimise for SC. Pls share feedback, thanks!
5. Sure.
6. Made corrections and updated.

File: Week-6/Assignment6/main.js [hunk 2]
```diff
@@ -27,3 +28,76 @@ console.log(sum3([-6,5,5,4],5)); //output: 4
 console.log(sum3([-9,1,8,-5],6)); //output: 4 
 console.log(sum3([4,0,5,-5,3,3,0,-4,-5],-2)); //output: -2  
 
+//approach2
+const threeSum = function(arr,target) {
+    // to address edge cases where input array is empty, null or undefined
+    if(!arr?.length) return `Array not acceptable`;
```
Comment (imranmd) [inline Week-6/Assignment6/main.js:34]: Perfect

File: Week-6/Assignment6/main.js [hunk 3]
```diff
@@ -27,3 +28,76 @@ console.log(sum3([-6,5,5,4],5)); //output: 4
...
+    while(left?.length) {
+        sortedArr.push(left.shift());
+    }
+    // copy the remaining elements of right[], if there are any
+    while(right?.length){
+        sortedArr.push(right.shift());
+    }
+    return sortedArr;
+} 
+//TC: O(N^2)
```
Comment (imranmd) [inline Week-6/Assignment6/main.js:97]: Is the TC Correct ?

**Review by imranmd — APPROVED** [review]
Re-evaluate Time complexity 
Approved with suggestions

**Pull request merged**